File: pkg/safast/fountain.go
package safast

import (
	"errors"
	"math"
)

const (
	ln2          = 0.6931471805599453
	solitonC     = 0.1
	solitonDelta = 0.5
)

func deterministicLog(x float64) float64 {
	e, m := 0, x
	for m >= 1.5 {
		m /= 2
		e++
	}
	for m < 0.75 {
		m *= 2
		e--
	}
	z := (m - 1) / (m + 1)
	z2 := z * z
	term, sum := z, 0.0
	for n := 1; n <= 21; n += 2 {
		sum += term / float64(n)
		term *= z2
	}
	return float64(e)*ln2 + 2*sum
}

func solitonCDF(k int) []float64 {
	cdf := make([]float64, k)
	if k == 1 {
		cdf[0] = 1
		return cdf
	}
	r := math.Max(1, solitonC*deterministicLog(float64(k)/solitonDelta)*math.Sqrt(float64(k)))
	spike := int(math.Ceil(float64(k) / r))
	if spike > k {
		spike = k
	}
	total := 0.0
	for d := 1; d <= k; d++ {
		rho := 1.0 / float64(d*(d-1))
		if d == 1 {
			rho = 1.0 / float64(k)
		}
		tau := 0.0
		if d < spike {
			tau = r / float64(d*k)
		} else if d == spike {
			tau = (r * math.Max(0, deterministicLog(r/solitonDelta))) / float64(k)
		}
		total += rho + tau
		cdf[d-1] = total
	}
	for i := range cdf {
		cdf[i] /= total
	}
	cdf[k-1] = 1
	return cdf
}

type splitmix32 struct{ state uint32 }

func (s *splitmix32) next() uint32 {
	s.state += 0x9e3779b9
	t := s.state ^ (s.state >> 16)
	t *= 0x21f0aaad
	t ^= t >> 15
	t *= 0x735a2d97
	t ^= t >> 15
	return t
}

func frameSeed(sessionID uint16, seq uint32) uint32 {
	h := uint32(sessionID+1)*0x9e3779b1 ^ (seq + 0x85ebca6b)
	h = (h ^ (h >> 13)) * 0xc2b2ae35
	return h ^ (h >> 16)
}

func frameIndices(k int, cdf []float64, sessionID uint16, seq uint32) []int {
	rng := splitmix32{state: frameSeed(sessionID, seq)}
	u := float64(rng.next()) * math.Ldexp(1, -32)
	lo, hi := 0, k-1
	for lo < hi {
		mid := (lo + hi) >> 1
		if cdf[mid] >= u {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	degree := lo + 1
	if degree > k {
		degree = k
	}
	if degree > k>>3 {
		scratch := make([]int, k)
		for i := range scratch {
			scratch[i] = i
		}
		out := make([]int, degree)
		for i := 0; i < degree; i++ {
			j := i + int(rng.next()%uint32(k-i))
			scratch[i], scratch[j] = scratch[j], scratch[i]
			out[i] = scratch[i]
		}
		return out
	}
	set := make(map[int]struct{}, degree)
	for len(set) < degree {
		set[int(rng.next()%uint32(k))] = struct{}{}
	}
	out := make([]int, 0, degree)
	for i := range set {
		out = append(out, i)
	}
	return out
}

type LTEncoder struct {
	K         int
	BlockLen  int
	SessionID uint16
	blocks    [][]byte
	cdf       []float64
}

func NewLTEncoder(payload []byte, blockLen int, sessionID uint16) (*LTEncoder, error) {
	if len(payload) == 0 || blockLen <= 0 {
		return nil, errors.New("payload and block length must be non-zero")
	}
	k := (len(payload) + blockLen - 1) / blockLen
	if k > 0xffff {
		return nil, errors.New("source block count exceeds uint16 wire limit")
	}
	blocks := make([][]byte, k)
	for i := 0; i < k; i++ {
		blocks[i] = make([]byte, blockLen)
		start := i * blockLen
		end := start + blockLen
		if end > len(payload) {
			end = len(payload)
		}
		copy(blocks[i], payload[start:end])
	}
	return &LTEncoder{K: k, BlockLen: blockLen, SessionID: sessionID, blocks: blocks, cdf: solitonCDF(k)}, nil
}

func (e *LTEncoder) Encode(seq uint32) []byte {
	out := make([]byte, e.BlockLen)
	for _, idx := range frameIndices(e.K, e.cdf, e.SessionID, seq) {
		for i, b := range e.blocks[idx] {
			out[i] ^= b
		}
	}
	return out
}

type pendingFrame struct {
	indices map[int]struct{}
	bytes   []byte
}

type LTDecoder struct {
	K         int
	BlockLen  int
	SessionID uint16
	TotalLen  int
	cdf       []float64
	solved    [][]byte
	byBlock   map[int]map[*pendingFrame]struct{}
	seen      map[uint32]struct{}
	Solved    int
	FramesNew int
	FramesDup int
}

func NewLTDecoder(k, blockLen int, sessionID uint16, totalLen int) (*LTDecoder, error) {
	if k <= 0 || k > 0xffff || blockLen <= 0 || totalLen <= 0 {
		return nil, errors.New("invalid decoder dimensions")
	}
	return &LTDecoder{K: k, BlockLen: blockLen, SessionID: sessionID, TotalLen: totalLen, cdf: solitonCDF(k), solved: make([][]byte, k), byBlock: map[int]map[*pendingFrame]struct{}{}, seen: map[uint32]struct{}{}}, nil
}

func xorInto(dst, src []byte) {
	for i := range dst {
		dst[i] ^= src[i]
	}
}

func (d *LTDecoder) Complete() bool { return d.Solved >= d.K }

func (d *LTDecoder) AddFrame(seq uint32, block []byte) error {
	if len(block) != d.BlockLen {
		return ErrWrongLength
	}
	if _, ok := d.seen[seq]; ok {
		d.FramesDup++
		return nil
	}
	d.seen[seq] = struct{}{}
	d.FramesNew++
	if d.Complete() {
		return nil
	}
	indices := map[int]struct{}{}
	for _, idx := range frameIndices(d.K, d.cdf, d.SessionID, seq) {
		indices[idx] = struct{}{}
	}
	data := append([]byte(nil), block...)
	for idx := range indices {
		if solved := d.solved[idx]; solved != nil {
			xorInto(data, solved)
			delete(indices, idx)
		}
	}
	if len(indices) == 0 {
		return nil
	}
	if len(indices) == 1 {
		for idx := range indices {
			d.resolve(idx, data)
		}
		return nil
	}
	pf := &pendingFrame{indices: indices, bytes: data}
	for idx := range indices {
		if d.byBlock[idx] == nil {
			d.byBlock[idx] = map[*pendingFrame]struct{}{}
		}
		d.byBlock[idx][pf] = struct{}{}
	}
	return nil
}

func (d *LTDecoder) resolve(first int, firstBytes []byte) {
	type item struct {
		idx   int
		bytes []byte
	}
	queue := []item{{first, firstBytes}}
	for len(queue) > 0 {
		last := len(queue) - 1
		cur := queue[last]
		queue = queue[:last]
		if d.solved[cur.idx] != nil {
			continue
		}
		d.solved[cur.idx] = append([]byte(nil), cur.bytes...)
		d.Solved++
		waiting := d.byBlock[cur.idx]
		delete(d.byBlock, cur.idx)
		for pf := range waiting {
			xorInto(pf.bytes, cur.bytes)
			delete(pf.indices, cur.idx)
			if len(pf.indices) == 1 {
				for next := range pf.indices {
					delete(d.byBlock[next], pf)
					if d.solved[next] == nil {
						queue = append(queue, item{next, pf.bytes})
					}
				}
			}
		}
	}
}

func (d *LTDecoder) Assemble() ([]byte, bool) {
	if !d.Complete() {
		return nil, false
	}
	out := make([]byte, d.TotalLen)
	for i, block := range d.solved {
		start := i * d.BlockLen
		if start >= len(out) {
			break
		}
		end := start + d.BlockLen
		if end > len(out) {
			end = len(out)
		}
		copy(out[start:end], block[:end-start])
	}
	return out, true
}

File: python/src/safast/fountain.py
from __future__ import annotations

from dataclasses import dataclass, field
import math

_LN2 = 0.6931471805599453
_SOLITON_C = 0.1
_SOLITON_DELTA = 0.5
_MASK32 = 0xFFFFFFFF


def _dlog(value: float) -> float:
    exponent = 0
    mantissa = value
    while mantissa >= 1.5:
        mantissa /= 2.0
        exponent += 1
    while mantissa < 0.75:
        mantissa *= 2.0
        exponent -= 1
    z = (mantissa - 1.0) / (mantissa + 1.0)
    z2 = z * z
    term = z
    total = 0.0
    for n in range(1, 22, 2):
        total += term / n
        term *= z2
    return exponent * _LN2 + 2.0 * total


def _soliton_cdf(k: int) -> list[float]:
    if k <= 0:
        raise ValueError("k must be positive")
    if k == 1:
        return [1.0]
    r = max(1.0, _SOLITON_C * _dlog(k / _SOLITON_DELTA) * math.sqrt(k))
    spike = min(k, math.ceil(k / r))
    cdf: list[float] = []
    total = 0.0
    for degree in range(1, k + 1):
        rho = 1 / k if degree == 1 else 1 / (degree * (degree - 1))
        tau = 0.0
        if degree < spike:
            tau = r / (degree * k)
        elif degree == spike:
            tau = (r * max(0.0, _dlog(r / _SOLITON_DELTA))) / k
        total += rho + tau
        cdf.append(total)
    cdf = [value / total for value in cdf]
    cdf[-1] = 1.0
    return cdf


class _SplitMix32:
    def __init__(self, seed: int) -> None:
        self.state = seed & _MASK32

    def next(self) -> int:
        self.state = (self.state + 0x9E3779B9) & _MASK32
        value = self.state ^ (self.state >> 16)
        value = (value * 0x21F0AAAD) & _MASK32
        value ^= value >> 15
        value = (value * 0x735A2D97) & _MASK32
        value ^= value >> 15
        return value & _MASK32


def _frame_seed(session_id: int, sequence: int) -> int:
    value = (((session_id + 1) * 0x9E3779B1) & _MASK32) ^ ((sequence + 0x85EBCA6B) & _MASK32)
    value = ((value ^ (value >> 13)) * 0xC2B2AE35) & _MASK32
    return (value ^ (value >> 16)) & _MASK32


def _frame_indices(k: int, cdf: list[float], session_id: int, sequence: int) -> list[int]:
    rng = _SplitMix32(_frame_seed(session_id, sequence))
    sample = rng.next() * 2.0**-32
    low, high = 0, k - 1
    while low < high:
        middle = (low + high) >> 1
        if cdf[middle] >= sample:
            high = middle
        else:
            low = middle + 1
    degree = min(k, low + 1)
    if degree > (k >> 3):
        scratch = list(range(k))
        output: list[int] = []
        for index in range(degree):
            swap = index + rng.next() % (k - index)
            scratch[index], scratch[swap] = scratch[swap], scratch[index]
            output.append(scratch[index])
        return output
    selected: set[int] = set()
    while len(selected) < degree:
        selected.add(rng.next() % k)
    return list(selected)


def _xor_into(destination: bytearray, source: bytes | bytearray) -> None:
    for index, value in enumerate(source):
        destination[index] ^= value


class LTEncoder:
    def __init__(self, payload: bytes, block_len: int, session_id: int) -> None:
        if not payload or block_len <= 0:
            raise ValueError("payload and block_len must be non-zero")
        self.block_len = block_len
        self.session_id = session_id
        self.k = max(1, math.ceil(len(payload) / block_len))
        if self.k > 0xFFFF:
            raise ValueError("source block count exceeds uint16 wire limit")
        self._blocks = []
        for index in range(self.k):
            block = payload[index * block_len : (index + 1) * block_len]
            self._blocks.append(block.ljust(block_len, b"\0"))
        self._cdf = _soliton_cdf(self.k)

    def encode(self, sequence: int) -> bytes:
        output = bytearray(self.block_len)
        for index in _frame_indices(self.k, self._cdf, self.session_id, sequence):
            _xor_into(output, self._blocks[index])
        return bytes(output)


@dataclass(eq=False, slots=True)
class _PendingFrame:
    indices: set[int]
    data: bytearray


@dataclass(slots=True)
class LTDecoder:
    k: int
    block_len: int
    session_id: int
    total_len: int
    solved_count: int = 0
    frames_new: int = 0
    frames_duplicate: int = 0
    _cdf: list[float] = field(init=False, repr=False)
    _solved: list[bytes | None] = field(init=False, repr=False)
    _by_block: dict[int, set[_PendingFrame]] = field(init=False, repr=False)
    _seen: set[int] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        if self.k <= 0 or self.k > 0xFFFF or self.block_len <= 0 or self.total_len <= 0:
            raise ValueError("invalid decoder dimensions")
        self._cdf = _soliton_cdf(self.k)
        self._solved = [None] * self.k
        self._by_block = {}
        self._seen = set()

    @property
    def complete(self) -> bool:
        return self.solved_count >= self.k

    def add_frame(self, sequence: int, block: bytes) -> None:
        if len(block) != self.block_len:
            raise ValueError("block length mismatch")
        if sequence in self._seen:
            self.frames_duplicate += 1
            return
        self._seen.add(sequence)
        self.frames_new += 1
        if self.complete:
            return
        indices = set(_frame_indices(self.k, self._cdf, self.session_id, sequence))
        data = bytearray(block)
        for index in tuple(indices):
            solved = self._solved[index]
            if solved is not None:
                _xor_into(data, solved)
                indices.remove(index)
        if not indices:
            return
        if len(indices) == 1:
            self._resolve(next(iter(indices)), data)
            return
        pending = _PendingFrame(indices, data)
        for index in indices:
            self._by_block.setdefault(index, set()).add(pending)

    def _resolve(self, first_index: int, first_data: bytearray) -> None:
        queue: list[tuple[int, bytearray]] = [(first_index, first_data)]
        while queue:
            index, data = queue.pop()
            if self._solved[index] is not None:
                continue
            self._solved[index] = bytes(data)
            self.solved_count += 1
            waiting = self._by_block.pop(index, set())
            for pending in waiting:
                _xor_into(pending.data, data)
                pending.indices.discard(index)
                if len(pending.indices) == 1:
                    next_index = next(iter(pending.indices))
                    self._by_block.get(next_index, set()).discard(pending)
                    if self._solved[next_index] is None:
                        queue.append((next_index, pending.data))

    def assemble(self) -> bytes | None:
        if not self.complete:
            return None
        return b"".join(block for block in self._solved if block is not None)[: self.total_len]
